# IMDB Movie Review Sentiment | Sequence Classification with LSTM Recurrent Neural Networks

In [23]:
from keras.datasets import imdb

In [24]:
import numpy 
from keras.datasets import imdb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence

In [25]:
top_words = 5000
(X_train, y_train), (X_test, y_test) = imdb.load_data(nb_words=top_words)
Max_review_len = 500
X_train= sequence.pad_sequences(X_train,maxlen=Max_review_len)
X_test= sequence.pad_sequences(X_test,maxlen=Max_review_len)

/Users/animeshkansal/anaconda/lib/python3.6/site-packages/keras/datasets/imdb.py:49: UserWarning: The `nb_words` argument in `load_data` has been renamed `num_words`.
  warnings.warn('The `nb_words` argument in `load_data` '


# Simple LSTM

In [35]:
embedding_vector_length = 32
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(LSTM(100))
model.add(Dense(1,activation='sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam', metrics=['accuracy'])

In [36]:
print(model.summary())
model.fit(X_train,y_train,epochs = 4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_9 (Embedding)      (None, 500, 32)           160000    
_________________________________________________________________
lstm_7 (LSTM)                (None, 100)               53200     
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 101       
Total params: 213,301
Trainable params: 213,301
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/4
25000/25000 [==============================] - 195s 8ms/step - loss: 0.5110 - acc: 0.7380

# Evaluating on Test Set

In [37]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  87.432 %


# But, LSTM generally have the problem of overfitting.



# Hence, Solution 1
# Adding Dropout layers between the Embedding and LSTM layers and the LSTM and Dense output layers.

In [38]:
from keras.layers import Dropout

In [39]:
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(Dropout(0.2))
model.add(LSTM(100))
model.add(Dropout(0.2))
model.add(Dense(1,activation = 'sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam', metrics=['accuracy'])


In [40]:
print(model.summary())
model.fit(X_train,y_train,epochs = 4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_10 (Embedding)     (None, 500, 32)           160000    
_________________________________________________________________
dropout_3 (Dropout)          (None, 500, 32)           0         
_________________________________________________________________
lstm_8 (LSTM)                (None, 100)               53200     
_________________________________________________________________
dropout_4 (Dropout)          (None, 100)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 101       
Total params: 213,301
Trainable params: 213,301
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/4
25000/25000 [==============================] - 232s 9ms/step - loss: 0.4842 - acc: 0.7660

# Evaluating on Test Set

In [41]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  87.184 %


# Solution 2

# Alternately, dropout can be applied to the input and recurrent connections of the memory units with the LSTM

In [44]:
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(LSTM(100, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1,activation='sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam', metrics=['accuracy'])


In [45]:
print(model.summary())
model.fit(X_train,y_train,epochs = 4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_12 (Embedding)     (None, 500, 32)           160000    
_________________________________________________________________
lstm_10 (LSTM)               (None, 100)               53200     
_________________________________________________________________
dense_10 (Dense)             (None, 1)                 101       
Total params: 213,301
Trainable params: 213,301
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/4
25000/25000 [==============================] - 237s 9ms/step - loss: 0.4868 - acc: 0.7646

# Evaluating on Test Set

In [46]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  77.948 %


# LSTM and Convolutional Neural Network For Sequence Classification

In [48]:
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D

In [51]:
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(Conv1D(filters=32,kernel_size=3,padding="same",activation='relu'))
model.add(MaxPooling1D(pool_size=2))
model.add(LSTM(100))
model.add(Dense(1,activation='sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

In [52]:
print(model.summary())
model.fit(X_train,y_train,epochs=4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_15 (Embedding)     (None, 500, 32)           160000    
_________________________________________________________________
conv1d_2 (Conv1D)            (None, 500, 32)           3104      
_________________________________________________________________
max_pooling1d_2 (MaxPooling1 (None, 250, 32)           0         
_________________________________________________________________
lstm_12 (LSTM)               (None, 100)               53200     
_________________________________________________________________
dense_12 (Dense)             (None, 1)                 101       
Total params: 216,405
Trainable params: 216,405
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/4
25000/25000 [==============================] - 123s 5ms/step - loss: 0.4205 - acc: 0.7993

# Evaluating on Test Set

In [53]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  88.1 %


# Adding Dropout layers between the Embedding and LSTM layers and the LSTM and Dense output layers.

In [61]:
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(Conv1D(filters=32,kernel_size=3,padding="same",activation='relu'))
model.add(MaxPooling1D(pool_size=2))
model.add(Dropout(0.2))
model.add(LSTM(100))
model.add(Dropout(0.2))
model.add(Dense(1,activation = 'sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam', metrics=['accuracy'])


In [62]:
print(model.summary())
model.fit(X_train,y_train,epochs=4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_18 (Embedding)     (None, 500, 32)           160000    
_________________________________________________________________
conv1d_5 (Conv1D)            (None, 500, 32)           3104      
_________________________________________________________________
max_pooling1d_5 (MaxPooling1 (None, 250, 32)           0         
_________________________________________________________________
dropout_7 (Dropout)          (None, 250, 32)           0         
_________________________________________________________________
lstm_15 (LSTM)               (None, 100)               53200     
_________________________________________________________________
dropout_8 (Dropout)          (None, 100)               0         
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 101       
Total para

# Evaluating on Test Set

In [63]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  86.488 %


# Alternately, dropout can be applied to the input and recurrent connections of the memory units with the LSTM

In [57]:
model = Sequential()
model.add(Embedding(top_words,embedding_vector_length,input_length=Max_review_len))
model.add(Conv1D(filters=32,kernel_size=3,padding="same",activation='relu'))
model.add(MaxPooling1D(pool_size=2))
model.add(LSTM(100, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1,activation = 'sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='adam', metrics=['accuracy'])

In [58]:
print(model.summary())
model.fit(X_train,y_train,epochs=4,batch_size=64)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_17 (Embedding)     (None, 500, 32)           160000    
_________________________________________________________________
conv1d_4 (Conv1D)            (None, 500, 32)           3104      
_________________________________________________________________
max_pooling1d_4 (MaxPooling1 (None, 250, 32)           0         
_________________________________________________________________
lstm_14 (LSTM)               (None, 100)               53200     
_________________________________________________________________
dense_14 (Dense)             (None, 1)                 101       
Total params: 216,405
Trainable params: 216,405
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/4
25000/25000 [==============================] - 161s 6ms/step - loss: 0.5008 - acc: 0.7470

# Evaluating on Test Set

In [60]:
scores = model.evaluate(X_test,y_test,verbose = 0 )
print("Accuracy = ",scores[1]*100,"%")

Accuracy =  85.368 %


# Final Observations

1. Simple LSTM Training Set Accuracy = 91.22%
2. Simple LSTM Test Set Accuracy = 87.43%

3. LSTM + Dropout Train Set Accuracy = 91.06%
4. LSTM + Dropout Test Set Accuracy = 87.18%

5. LSTM + Dropout_2 Train Set Accuracy = 84.99%
6. LSTM + Dropout_2 Test Set Accuracy = 77.95%

7. CNN + LSTM Train Set Accuracy = 93.84%
8. CNN + LSTM Test Set Accuracy = 88.1%

9. CNN + LSTM + Dropout Train Set Accuracy = 93.57%
10. CNN + LSTM + Dropout Test Set Accuracy = 86.49%

11. CNN + LSTM + Dropout_2 Train Set Accuracy = 90.50%
12. CNN + LSTM + Dropout_2 Test Set Accuracy = 85.37%
 

# I have run all the models for 4 epochs , because each epoch take 3-4 minutes to run,  Number of Epochs will change the result .

# As per current situation : The Algorithm that worked best on test set is : 

# CNN + LSTM without dropout